Confirm receiving merge commands

duijf opened the pull request:
Fixes https://github.com/channable/hoff/issues/11

This makes Hoff post a comment that confirms it has received a valid merge command. This makes it more user-friendly to see when PRs have been added to the merge queue.

ruuda commented on src/Logic.hs:
```diff
@@ -272,16 +272,19 @@ handleCommentAdded
+    isValidApproveCommand = (isMergeCommand triggerConfig body) && (isReviewer projectConfig author)
   in
-    pure . Pr.updatePullRequest pr update
+    if isValidApproveCommand
```
This will conflict with #20, because `isReviewer` is needs to run in the monad. But it should be easy to fix.

ruuda commented on src/Logic.hs:
```diff
@@ -272,16 +272,19 @@ handleCommentAdded
-    pure . Pr.updatePullRequest pr update
+    if isValidApproveCommand
+      then do
+        leaveComment pr "Added PR to merge queue"
```
I was thinking we could add who approved it an what its position in the queue to the comment too. We can inspect the queue with `filterPullRequestsBy` to filter on approved-but-not-conflicted PRs, and then filter the remaining ids on `< pr`, to count how many there are before this one.

duijf commented:
I rebased this PR onto your changes now. 

Tomorrow, I'll add a test and will change the message about the queue position.

duijf commented:
@ruuda The confirmation message now mentions who approved the PR and provides some info about the queue status.

One thing that users might not find intuitive is that PRs are currently merged by lowest ID instead of by approval order. The message `Rebasing after 1 other PR(s)` is not completely accurate. 

I'd be open to changing the message to something more accurate. (Or we could make Hoff rebase PRs in the order they were approved).

wesleybowman commented:
> (Or we could make Hoff rebase PRs in the order they were approved).

Wouldn't this be better in general anyway? What if I open a PR, and then a reviewer is like, "Please split out the typing from this PR in a separate PR..." or something similar, and then the new PR is approved as well as the old. It seems like you should be able to request a merge in the order you want, so I could go to the new PR and request a merge, and then go to the older PR and request immediately as well, without having to wait on the newer one to be merged first.

duijf commented:
> Wouldn't this be better in general anyway?

I think so too. I'll create a separate issue for this so we can discuss in more detail.

ruuda commented on src/Logic.hs:
```diff
@@ -284,10 +284,22 @@ handleCommentAdded triggerConfig pr author body state = do
+        queueLength = Pr.getQueueLength pr newState
+        queueInfo = if queueLength == 0
+          then "Merge queue is empty. Rebasing immediately."
+          else format "Rebasing after {} other PR(s)." [queueLength]
```
It is possible to match on integers. So we can do something like

```haskell
case Pr.getQueueLength pr newState of
  0 -> format "Pull request approved by @{}, rebasing now." [author]
  1 -> format "Pull request approved by @{}, waiting for rebase at the front of the queue." [author]
  n -> format "Pull request approved by @{}, waiting for rebase behind {} pull requests." [author, n]
```

ruuda commented on tests/Spec.hs:
```diff
@@ -234,7 +234,27 @@ main = hspec $ do
+    it "notifies approvers about queue position" $ do
+      let
+        state0 = candidateState (PullRequestId 1) (Branch "p") (Sha "a38") "tyrell" (Sha "84c")
+        state1 = Project.insertPullRequest (PullRequestId 2) (Branch "s") (Sha "dec") "Some PR" (Username "rachel") state0
```
(It’s [Rach**a**el](https://en.wikipedia.org/wiki/List_of_Blade_Runner_characters#Rachael), not Rachel.)

duijf replied:
I was wondering what this was referring to!

ruuda commented on tests/Spec.hs:
```diff
@@ -234,7 +234,27 @@ main = hspec $ do
+        state1 = Project.insertPullRequest (PullRequestId 2) (Branch "s") (Sha "dec") "Some PR" (Username "rachel") state0
+        event0 = CommentAdded (PullRequestId 1) "deckard" "@bot merge"
+        event1 = CommentAdded (PullRequestId 2) "deckard" "@bot merge"
+        (state2, actions0) = handleEventFlat event0 state1
```
There is also `handleEventsFlat` that takes a list of events.

duijf commented:
Thanks for the suggestions, they indeed make the code a lot easier to read! I believe I addressed all your comments.

I've rebased on `master` and fixed up the history. NB: GitHub seems to re-order the commits by date, but the actual order in the graph is:

```
927b14e (HEAD -> confirm-merge-command) github: Confirm receiving merge commands
0ee76ca project: Add function to calculate queue length
ad0115a test: Hoff confirms merge commands
a4856e2 (origin/master) Merge #29
```

ruuda commented:
> NB: GitHub seems to re-order the commits by date

It does, this was one of the reasons that having to put the commit hash in the LGTM stamp to trigger a merge was inconvenient, because the last one on GitHub is not necessarily the commit you want to merge :confused:

ruuda commented on tests/Spec.hs:
```diff
@@ -244,7 +245,29 @@ main = hspec $ do
+      let
+        state = candidateState (PullRequestId 1) (Branch "p") (Sha "a38") "tyrell" (Sha "84c")
+              & Project.insertPullRequest (PullRequestId 2) (Branch "s") (Sha "dec") "Some PR" (Username "rachael")
+              & Project.insertPullRequest (PullRequestId 3) (Branch "s") (Sha "f16") "Another PR" (Username "raechael")
```
Rachael, not Raechael :see_no_evil:

duijf replied:
Lol damnit :see_no_evil:

duijf commented:
@OpsBotPrime merge

OpsBotPrime commented:
Rebased as 67702a3497d8fe32969a09113b8c5cb2b524f853, waiting for CI …
